types/connector.d.ts

```diff
@@ -3,28 +3,285 @@
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
 
-import {Callback, DataSource, Options, PromiseOrVoid} from '..';
+import { AnyObject } from 'strong-globalize/lib/config';
+import {Callback, DataSource, Filter, ModelBase, ModelBaseClass, ModelDefinition, ModelProperties, Options, PromiseOrVoid, PropertyDefinition, PropertyType, Schema, Where} from '..';
 
-// Copyright IBM Corp. 2018. All Rights Reserved.
-// Node module: loopback-datasource-juggler
-// This file is licensed under the MIT License.
-// License text available at https://opensource.org/licenses/MIT
+export type ConnectorSettings = Options & {
+  name?: string,
+  /**
+   * Overrides {@link ConnectorSettings.adapter} if defined.
+   */
+  connector?: ConnectorStatic | string,
+  /**
+   * @deprecated Use {@link ConnectorSettings.connector} instead.
+   */
+  adapter?: ConnectorStatic | string,
+  connectionTimeout?: number,
+  maxOfflineRequests?: number,
+  lazyConnect?: boolean,
+  debug?: boolean,
+}
+
+export type IDPropertiesDiscoveryOptions = {
+  owner?: string,
+} | {
+  schema?: string
+}
+
+export type DiscoveryScopeOptions = {
+  owner?: string,
+  all?: boolean,
+  views?: boolean,
+  limit?: number,
+  offset?: number,
+}
+
+export type SchemaDiscoveryOptions = {
+  owner?: string,
+  /**
+   * Overriden by {@link SchemaDiscoveryOptions.associations}.
+   */
+  relations?: boolean,
+  all?: boolean,
+  views?: boolean,
+  disableCamelCase?: boolean,
+  nameMapper?: NameMapper,
+  associations?: boolean,
+}
+
+export type NameMapper = (type: 'table' | 'model' | 'fk' | string, name: string) => string | null;
+
+export type DiscoveredPrimaryKeys = {
+  owner: string | null,
+  tableName: string,
+  columnName: string,
+  keySeq: number,
+  pkName: string | null,
+}
+
+export type DiscoveredForeignKeys = {
+  fkOwner: string | null,
+  fkName: string | null,
+  fkTableName: string,
+  fkColumnName: string,
+  keySeq: number,
+
+  pkOwner: string | null,
+  pkName: string | null,
+  pkTableName: string,
+  pkColumnName: string,
+}
+
+export type DiscoveredModelProperties = {
+  owner?: string,
+  tableName?: string,
+  columnName?: string,
+  dataType?: string,
+  dataLength?: number,
+  dataPrecision?: number,
+  dataScale?: number,
+  nullable?: boolean,
+}
+
+
+// #TODO(achrinza): The codebase suggets that `context` differs
+// depending on the situation, and that there's no cohesive interface.
+// Hence, we'll need to identify all the possible contexts.
+export type Context = {
+  Model: ModelBaseClass,
+  instance?: object,
+  query?: Filter,
+  where?: Where,
+  data?: AnyObject,
+  hookState?: AnyObject,
+  options?: Options,
+  isNewInstance?: boolean,
+  currentInstance?: Readonly<ModelBase>,
+}
```

Comment:
These are defined here:

- https://loopback.io/doc/en/lb3/Operation-hooks.html
- https://loopback.io/doc/en/lb3/Remote-hooks.html#context-object
- https://loopback.io/doc/en/lb3/Connector-hooks.html#context

Currently, this is only based on all the possible attributes for an Operation Hooks context.

Will probably need to add separate context typedefs for the other 2.

While the Operation Hooks context differs depending on the event being observed and the function called, I'm not sure if its better to create distinctive op. hook contexts for these different hooks.